### ISSUE TITLE
Epic - Performance and scale issues in 1.0

### ISSUE THREAD

**Issue by louiscryan** [istio/istio#7879]
Using this issue to track and provide overall status for the primary performance and scaling issues in 1.0.

Team will use this to organize itself and resolution of these issues

**Comment by louiscryan** [comment]
From Envoy too

https://github.com/envoyproxy/envoy/pull/4117

**Comment by mandarjog** [comment]
You mention p99 latecies shot up as pods scaled? Which pods were you talking about ? Ingress gateway or application pods?

**Comment by none-da** [comment]
@mandarjog istio pods. Ingress gateway, mixer and pilot pods.

**Comment by mandarjog** [comment]
For 1.0.2 (1.0.3-pre) our results are different.

1. Were any of the pods cpu constrained ?
2. For latency sensitive operations you should at least request 1 full cpu per proxy.
3. Remove `stdio` and `stdiotcp` rules from mixer
4. disable access logging at the proxy

**Comment by jaygorrell** [comment]
>For 1.0.2 (1.0.3-pre) our results are different.

Do you have results using large pod counts or just high traffic on the 2 pod test? In a small pod-count environment I've had no issues but as counts increase to upwards of 900 or so, Istio system pods seem to struggle, which bubbles up into all kinds of issues.

I believe the bottlenecks appear to be on the pilot's discovery pod and the telemetry pods for that though. 1.1.0 had a big improvement but the problem isn't gone there either.

**Comment by mandarjog** [comment]
https://grafana.v103.qualistio.org/d/1/istio-mesh-dashboard?refresh=5s&orgId=1

This `1.0.3-pre` is running with 400 services and a total of 800 pods.

For this size and mesh traffic at 60k rps, istio-telemetry needs 40-50 vCPU.
1 vcpu should be allocated to proxy.

https://kiali.v103.qualistio.org/console/graph/namespaces/all?layout=cose-bilkent&duration=60&edges=hide&graphType=versionedApp&injectServiceNodes=false

**Comment by jaygorrell** [comment]
That is very helpful, thank you!

Can you clarify a few things, please?
1. So with 800 pods here and 1vcpu for the proxy, you're saying this
environment has 800 vcpu plus the 40-50 for telemetry and then the services
themselves... so probably over 1000 vcpu?
2. This is 1.0.3-pre... I was not able to get similar results on 1.0.2 or
1.1.0-snapshot.1... are there explanatory changes in 1.0.3 for this or can
you get similar results on 1.1.0?
3. If the answer to #1 is yes, what options do users have to lower those
requirements? My cluster that has problems has 25 m4.xlarge nodes, which is
100 vcpu. It runs 800 pods just fine without istio but I haven't had any
luck after removing the stdout rules, turning off logging, and even playing
with the telemetry sampling size. I do hope I don't need to make the
cluster 10x as large here to 1000 vcpu.

**Comment by Stono** [comment]
Yeah that does sound like an awfully large overhead

**Comment by mandarjog** [comment]
Istio proxy cpu scales with how many total requests (internal and ingress) you are driving thru the system. 0.8 cpu per 1k rps is a good estimate. And 0.4 telemetry per 1k rps. 

Telemetry and pilot also scales with the number of sidecars that connect to it.

1.03, There is a perf improvement in telemetry that cuts down cpu by 10-15 %.

In your env 800 pods and 100 cpu, every pod has 0.125 cpu, but with low load that should be ok.

1.03 pilot also has perf improvements, specifically eds, which reduces pilot cpu in typical use case.

Bottom line, you do not need the telemetry cpus I quoted, but you should allocate 4 cpus to telemetry and then leave it to hpa.

**Comment by jaygorrell** [comment]
That's good info, thanks. 

The cluster I was referring to (#9043) only showed ~500 ops (not 500k), nearly all from liveness probes... but still had some issues. It was significantly better on 1.1.0 over 1.0.2 though... and I assume the 1.0.3 changes aren't in 1.1.0-snapshot.1? I think my problems on 1.1.0 were around individual hosts with telemetry running on them getting choked so I can probably do some node selections to run some of these components to keep that clear. 

1.0.2 didn't work at all for those pod counts though... it sounds promising that things are definitely improving!
